onnx/defs/training/defs.cc

```diff
@@ -324,4 +324,134 @@ ONNX_TRAINING_OPERATOR_SET_SCHEMA(
             OpSchema::all_tensor_types(),
             "Allow inputs and outputs to be any kind of tensor."));
 
+static const char* Adagrad_ver1_doc = R"DOC(
+    Compute one iteration of ADAGRAD, a stochastic gradient based optimization
+    algorithm. This operator can conduct the optimization of multiple tensor variables.
+
+    Let's define the behavior of this operator. As you can imagine, ADAGRAD requires
+    some parameters:
+     
+     - The initial learning-rate "R".
+     - The update count "T". That is, the number of training iterations conducted.
+     - A L2-norm regularization coefficient "norm_coefficient".
+     - A learning-rate decay factor "decay_factor".
+     - A small constant "epsilon" to avoid dividing-by-zero. 
+
+    At each ADAGRAD iteration, the optimized tensors are moved along a direction
+    computed based on their estimated gradient and accumulated squared gradient. Assume
+    that only a single tensor "X" is updated by this operator. We need the value of "X",
+    its gradient "G", and its accumulated squared gradient "H". Therefore, variables in
+    this operator's input list are sequentially "R", "T", "X", "G", and "H". Other
+    parameters are given as attributes because they are usually constants. Also, the
+    corresponding output tensors are the new value of "X" (called "X_new"), and then
+    the new accumulated squared gradient (called "H_new"). Those outputs are computed
+    from the given inputs following the pseudo code below.
+
+    Let "+", "-", "*", and "/" are all element-wise arithmetic operations with
+    numpy-style broadcasting support. The pseudo code to compute those outputs is:
+
+      // Compute a scalar learning-rate factor. If X is never updated, T should be 0.
+      r = R / (1 + T * decay_factor);
+
+      // Add gradient of 0.5 * norm_coefficient * ||X||_2^2, where ||X||_2 is the 2-norm.
+      G_regularized = norm_coefficient * X + G;
```

Comment:
The comment and pseudo-code above don't seem to match.

Reply:
It matches. Let `y=0.5 * coef * x^2`. We have `dy/dx = coef * x`.

Reply:
Sorry, I misunderstood the comment.